pkg/kv/kvtest/iterators.go

```diff
@@ -0,0 +1,425 @@
+			firstPartitionKey, []byte(""), []byte(""))
+		if err != nil {
+			t.Fatalf("failed to create secondary iterator: %v", err)
+		}
+		defer itr.Close()
+		names := make([]string, 0)
+		for itr.Next() {
+			e := itr.Entry()
+			model, ok := e.Value.(*TestModel)
+			if !ok {
+				t.Fatalf("cannot read from store")
+			}
+			names = append(names, string(model.Name))
+		}
+		if itr.Err() != nil {
+			t.Fatalf("unexpected error: %v", itr.Err())
+		}
+
+		if diffs := deep.Equal(names, []string{"a", "aa", "b", "c", "d"}); diffs != nil {
+			t.Fatalf("got wrong list of names: %v", diffs)
+		}
+	})
+
+	t.Run("listing with prefix", func(t *testing.T) {
+		itr, err := kv.NewSecondaryIterator(ctx, store, (&TestModel{}).ProtoReflect().Type(),
+			firstPartitionKey, []byte("a"), []byte(""))
+		if err != nil {
+			t.Fatalf("failed to create secondary iterator: %v", err)
+		}
+		defer itr.Close()
+		names := make([]string, 0)
+		for itr.Next() {
+			e := itr.Entry()
+			model, ok := e.Value.(*TestModel)
+			if !ok {
+				t.Fatalf("cannot read from store")
+			}
+			names = append(names, string(model.Name))
+		}
+		if itr.Err() != nil {
+			t.Fatalf("unexpected error: %v", itr.Err())
+		}
+
+		if diffs := deep.Equal(names, []string{"a", "aa"}); diffs != nil {
+			t.Fatalf("got wrong list of names: %v", diffs)
+		}
+	})
+
+	t.Run("listing after", func(t *testing.T) {
+		itr, err := kv.NewSecondaryIterator(ctx, store, (&TestModel{}).ProtoReflect().Type(),
+			firstPartitionKey, []byte(""), []byte("b"))
+		if err != nil {
+			t.Fatalf("failed to create secondary iterator: %v", err)
+		}
+		defer itr.Close()
+		names := make([]string, 0)
+		for itr.Next() {
+			e := itr.Entry()
+			model, ok := e.Value.(*TestModel)
+			if !ok {
+				t.Fatalf("cannot read from store")
+			}
+			names = append(names, string(model.Name))
+		}
+		if itr.Err() != nil {
+			t.Fatalf("unexpected error: %v", itr.Err())
+		}
+
+		if diffs := deep.Equal(names, []string{"c", "d"}); diffs != nil {
+			t.Fatalf("got wrong list of names: %v", diffs)
+		}
+	})
+
+	t.Run("key not found", func(t *testing.T) {
+		itr, err := kv.NewSecondaryIterator(ctx, store, (&TestModel{}).ProtoReflect().Type(),
+			thirdPartitionKey, []byte(""), []byte(""))
+		if err != nil {
+			t.Fatalf("failed to create secondary iterator: %v", err)
+		}
+		defer itr.Close()
+
+		for itr.Next() {
+			e := itr.Entry()
+			_, ok := e.Value.(*TestModel)
+			if !ok {
+				t.Fatalf("cannot read from store")
+			}
+		}
+
+		if !errors.Is(itr.Err(), kv.ErrMissingKey) {
+			t.Fatalf("expected error: %s, got %v", kv.ErrMissingKey, itr.Err())
+		}
+	})
+
+	t.Run("value as key not found", func(t *testing.T) {
+		itr, err := kv.NewSecondaryIterator(ctx, store, (&TestModel{}).ProtoReflect().Type(),
+			thirdPartitionKey, []byte(""), []byte("e"))
+		if err != nil {
+			t.Fatalf("failed to create secondary iterator: %v", err)
+		}
+		defer itr.Close()
+		names := make([]string, 0)
+		for itr.Next() {
```

Comment:
- should verify that next returns false.
- names diff can be removed as we not suppose to go into a loop here.
-